crates/ty_python_semantic/src/types.rs

```diff
@@ -4547,6 +4548,36 @@ impl<'db> Type<'db> {
             .to_instance(db)
             .expect("type should be convertible to instance type");
 
+        // Check if __new__ explicitly returns a different type than the instance
+        let new_return_type = new_call_outcome
+            .as_ref()
+            .and_then(|outcome| outcome.as_ref().ok())
+            .and_then(|bindings| {
+                let ret_type = bindings.return_type(db);
+
+                // Skip if return type is unknown/dynamic or None - likely unresolved annotations
+                if ret_type.is_unknown()
+                    || matches!(ret_type, Type::Dynamic(_))
```

Comment:
This is redundant since `Unknown` is a variant of `Dynamic`.

The [spec](https://typing.python.org/en/latest/spec/constructors.html#new-method) says that an explicit `Any` or a union containing `Any` should be considered as "not an instance". So I think that we should return `None` if the return type is `Unknown` (as "no annotation" should be considered same as `Self`), but if we see `Any` that should cause us to return `Some(ret_type)`.

We should also add some tests for these cases.